internal/btf/co_re.go

```diff
@@ -0,0 +1,737 @@
+	case reloTypeIDTarget:
+		return uint32(spec.rootType.ID()), nil
+	case reloTypeExists:
+		return 1, nil
+	case reloTypeSize:
+		sz, err := Sizeof(spec.rootType)
+		if err != nil {
+			return 0, err
+		}
+		return uint32(sz), nil
+	default:
+		return 0, errUnsupportedRelocation
+	}
+}
+
+func coreCalculateEnumvalRelocation(relo *CORERelocation, spec *coreSpec) (uint32, error) {
+	switch relo.Kind {
+	case reloEnumvalExists:
+		if spec != nil {
+			return 1, nil
+		} else {
+			return 0, nil
+		}
+	case reloEnumvalValue:
+		if spec == nil {
+			return 0, errStructureNeedsCleaning
+		}
+		acc := spec.spec[0]
+		t := acc.typ
+		et, ok := t.(*Enum)
+		if !ok {
+			return 0, fmt.Errorf("enumval relocation against non-enum type")
+		}
+		return uint32(et.Values[acc.idx].Value), nil
+	default:
+		return 0, errUnsupportedRelocation
+	}
+}
+
+func coreCalculateFieldRelocation(relo *CORERelocation, spec *coreSpec) (uint32, *bool, error) {
+	var val uint32
+	var validate bool
+
+	if relo.Kind == reloFieldExists {
+		if spec != nil {
+			return 1, nil, nil
+		} else {
+			return 0, nil, nil
+		}
+	}
+
+	if spec == nil {
+		return 0, nil, errStructureNeedsCleaning
+	}
+
+	acc := spec.spec[len(spec.spec)-1]
+	t := acc.typ
+
+	if acc.name == "" {
+		if relo.Kind == reloFieldByteOffset {
+			val = spec.bitOffset / 8
+		} else if relo.Kind == reloFieldByteSize {
+			sz, err := Sizeof(t)
+			if err != nil {
+				return 0, nil, err
+			}
+			val = uint32(sz)
+		} else {
+			return 0, nil, fmt.Errorf("relocation cannot be applied to array access")
+		}
+		validate = true
+		return val, &validate, nil
+	}
+
+	ct, ok := t.(composite)
+	if !ok {
+		return 0, nil, fmt.Errorf("non-composite type for relocation: %T", t)
+	}
+	m := ct.members()[acc.idx]
+	mt := skipModsAndTypedefs(m.Type)
+	bitOffset := spec.bitOffset
+	bitSize := m.BitfieldSize
+	var byteSize uint32
+	var byteOffset uint32
+
+	bitfield := bitSize > 0
+	if bitfield {
+		if ms, ok := mt.(sizer); ok {
+			byteSize = ms.size()
+			byteOff := bitOffset / 8 / byteSize * byteSize
+			for (bitOffset + bitSize - byteOff*8) > (byteSize * 8) {
+				if byteSize >= 8 {
+					return 0, nil, fmt.Errorf("cannot be satisfied for bitfield")
+				}
+				byteSize *= 2
+				byteOff = bitOffset / 8 / byteSize * byteSize
+			}
+		} else {
+			return 0, nil, fmt.Errorf("bit offset was non-zero for a non-sizeable type")
+		}
+	} else {
+		sz, err := Sizeof(m.Type)
+		if err != nil {
+			return 0, nil, err
+		}
+		byteSize = uint32(sz)
+		byteOffset = spec.bitOffset / 8
+		bitSize = byteSize / 8
+	}
+
+	validate = !bitfield
+
+	switch relo.Kind {
+	case reloFieldByteOffset:
+		val = byteOffset
+	case reloFieldByteSize:
+		val = byteSize
+	case reloFieldSigned:
+		switch v := mt.(type) {
+		case *Enum:
+			val = 1
+		case *Int:
+			if v.Encoding == Signed {
+				val = 1
+			} else {
+				val = 0
+			}
+		default:
+			val = 0
+		}
+		validate = true
+	case reloFieldLShiftU64:
+		if internal.NativeEndian == binary.LittleEndian {
```

Comment:
The shifts are only used for bitfield access. It reads the value, does a left shift, then a right shift in order to obtain only the bitfield you want. See https://github.com/libbpf/libbpf/commit/4438972ccc172ce2a68ee5d2ec2a01303aa380a8#diff-92cca15bc801d3458b089b6022edfcaaR49

I believe this depends on endianness because the bitfield will change order within a word depending on endianness, whereas the BTF representation does not.

Reply:
Ugh, what a mess. Can you add a comment that explains this please? To me it looked like shift left and shift right were general purpose constructs...

Reply:
Comment added.